casper-deploy-notifier/src/lib.rs

```diff
@@ -0,0 +1,97 @@
+use casper_types::AsymmetricType;
+use eventsource_stream::{Event, EventStreamError, Eventsource};
+use futures::stream::{BoxStream, TryStreamExt};
+use tokio::sync::mpsc;
+
+use crate::error::SseError;
+use crate::sse_types::SseData;
+use crate::types::Notification;
+
+mod error;
+mod sse_types;
+mod types;
+
+const DEFAULT_SSE_SERVER: &str = "https://events.mainnet.casperlabs.io";
+const DEFAULT_EVENT_CHANNEL: &str = "/events/main";
+
+type BoxedEventStream = BoxStream<'static, Result<Event, EventStreamError<reqwest::Error>>>;
+
+pub struct DeployNotifier {
+    url: String,
```

Comment:
use reqwest's URL type for more type safety

Reply:
Not possible if we want to provide `impl Default`.

Reply:
This just means that there should be no `Default` implementation for `DeployNotifier` how can a `DeployNotifier` listen to anything without knowing about a url. there should be no notifier if there is no URL it can listen to.

Reply:
There is no way to construct complex `reqwest::Url` structure - that unnecessarily exposes internal implementation - without throwing error, and that is reason to not provide `Default`? You think about strict Kairos usage, however from the beginning Casper Deploy Notifier was aimed to be easy to use library that could be shared with community. Most users are expected to rely on default URL instead of providing custom one, as there is lack of public infrastructure in Casper.

Reply:
You can call `parse` and unwrap if you are sure that the default URL you gave it is a valid URL, and still use `Default`. I think you misunderstand my comments. I'm trying to make suggestions towards a more simple API avoiding introducing state that is not obvious to the user.
As I said earlier:
```Rust
let notifier = default(); // is it listening yet, what does it listen to?
notifier.connect(); // what is it connecting to?
notifier.run(tx); // do I need to connect manually before or when it runs, do I need to disconnect manually?
```
```Rust
let notifier = new("http://test.com"); // is it listening yet?
notifier.run(tx); // do I need to connect manually before or when it runs, do I also need to disconnect manually?
```
```Rust
subscribe("http://test.com", sender); // no more work I need to do
```